# Data cleaning

### Importación de librerías

In [4]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

### Importación de datasets

In [58]:
df_players = pd.read_csv('datasets/players.csv')
fuente_players = "https://www.fide.com/"
df_2015 = pd.read_csv('datasets/ratings_2015.csv')
fuente_2015 = "https://www.fide.com/"
df_2016= pd.read_csv('datasets/ratings_2016.csv')
fuente_2016 = "https://www.fide.com/"
df_2017 = pd.read_csv('datasets/ratings_2017.csv')
fuente_2017 = "https://www.fide.com/"
df_2018 = pd.read_csv('datasets/ratings_2018.csv')
fuente_2018 = "https://www.fide.com/"
df_2019 = pd.read_csv('datasets/ratings_2019.csv')
fuente_2019 = "https://www.fide.com/"
df_2020 = pd.read_csv('datasets/ratings_2020.csv')
fuente_2020 = "https://www.fide.com/"
df_paises_ISO = pd.read_csv('datasets/paises.csv')
fuente_paises_ISO = "https://gist.github.com/kintero/7d1db891401f56256c79#file-paises-csv"

Con respecto a las distintas modalidades de juego que son válidas para Elo FIDE, se reconocen tres: estándar, rápido y blitz. Sin embargo, para los propósitos de este análisis, se ha restringido el enfoque únicamente a la modalidad estándar.

### Dataset 1: Clasificación de los jugadores en continentes y categorías de edad

In [59]:
df_players.head()

fide_id                  name federation gender title   yob
0   100013    Campora, Daniel H.        ARG      M    GM  1957
1   100021  Quinteros, Miguel A.        ARG      M    GM  1947
2   100048    Iapichello, Javier        ARG      M   NaN  1990
3   100064    Schipper, Bernardo        ARG      M   NaN  2000
4   100072           Adla, Diego        ARG      M    IM  1968

In [60]:
df_paises_ISO.head()

nombre         name          nom  iso2  iso3  phone_code continente
0  Afganistán  Afghanistan  Afghanistan    AF   AFG          93       Asia
1     Albania      Albania      Albanie    AL   ALB         355     Europa
2    Alemania      Germany    Allemagne    DE   DEU          49     Europa
3     Algeria      Algeria      Algérie    DZ   DZA         213     África
4     Andorra      Andorra      Andorra    AD   AND         376     Europa

In [61]:
df_2020

fide_id  year  month  rating_standard  rating_rapid  rating_blitz
0           100013  2020      1           2420.0        2350.0        2378.0
1           100021  2020      1           2422.0           NaN           NaN
2           100048  2020      1           1607.0           NaN           NaN
3           100064  2020      1           2116.0           NaN           NaN
4           100072  2020      1           2440.0           NaN           NaN
...            ...   ...    ...              ...           ...           ...
5125945  651079542  2020     12              NaN        1473.0           NaN
5125946  651080206  2020     12           1327.0           NaN           NaN
5125947  651081628  2020     12           1195.0           NaN           NaN
5125948  651081695  2020     12              NaN           NaN        1433.0
5125949  651081830  2020     12              NaN        1221.0           NaN

[5125950 rows x 6 columns]

In [62]:
df_paises_ISO_fed = df_paises_ISO.copy()
df_paises_ISO_fed = df_paises_ISO_fed.rename(columns={' iso3': 'federation'}) #quitamos el espacio inicial y cambiamos el nombre para que sea el mismo que en df_players
df_paises_ISO_fed = df_paises_ISO_fed.rename(columns={'continente': 'continent'}) #lo añadimos al ingles, como el resto de datasets (concordancia)
df_paises_ISO_fed = df_paises_ISO_fed[['federation', 'continent']] #columnas que interesan
df_paises_ISO_fed

federation            continent
0          AFG                 Asia
1          ALB               Europa
2          DEU               Europa
3          DZA               África
4          AND               Europa
..         ...                  ...
241        WLF  Australia y Oceanía
242        YEM                 Asia
243        DJI               África
244        ZMB               África
245        ZWE               África

[246 rows x 2 columns]

In [64]:
df_players_cont = pd.merge(df_players, df_paises_ISO_fed, on='federation', how='inner')
df_players_cont

fide_id                   name federation gender title   yob  \
0          100013     Campora, Daniel H.        ARG      M    GM  1957   
1          100021   Quinteros, Miguel A.        ARG      M    GM  1947   
2          100048     Iapichello, Javier        ARG      M   NaN  1990   
3          100064     Schipper, Bernardo        ARG      M   NaN  2000   
4          100072            Adla, Diego        ARG      M    IM  1968   
...           ...                    ...        ...    ...   ...   ...   
318996  651079542      Teisseire, Armand        FRA      M   NaN  2006   
318997  651080206         Lagarenne, Tom        FRA      M   NaN  1994   
318998  651081628        Kratz, Jonathan        FRA      M   NaN  1988   
318999  651081695          Bar, Francois        FRA      M   NaN  1965   
319000  651081830  Martinati, Jean Louis        FRA      M   NaN  1965   

       continent  
0        América  
1        América  
2        América  
3        América  
4        América  
...          ...  
318996    Europa  
318997    Europa  
318998    Europa  
318999    Europa  
319000    Europa  

[319001 rows x 7 columns]

A continuación añadimos la edad de cada jugador que tenían en 2020.

In [65]:
df_players_cont['years_id_2020'] = 2020 - df_players_cont.yob
df_players_cont

fide_id                   name federation gender title   yob  \
0          100013     Campora, Daniel H.        ARG      M    GM  1957   
1          100021   Quinteros, Miguel A.        ARG      M    GM  1947   
2          100048     Iapichello, Javier        ARG      M   NaN  1990   
3          100064     Schipper, Bernardo        ARG      M   NaN  2000   
4          100072            Adla, Diego        ARG      M    IM  1968   
...           ...                    ...        ...    ...   ...   ...   
318996  651079542      Teisseire, Armand        FRA      M   NaN  2006   
318997  651080206         Lagarenne, Tom        FRA      M   NaN  1994   
318998  651081628        Kratz, Jonathan        FRA      M   NaN  1988   
318999  651081695          Bar, Francois        FRA      M   NaN  1965   
319000  651081830  Martinati, Jean Louis        FRA      M   NaN  1965   

       continent  years_id_2020  
0        América             63  
1        América             73  
2        América             30  
3        América             20  
4        América             52  
...          ...            ...  
318996    Europa             14  
318997    Europa             26  
318998    Europa             32  
318999    Europa             55  
319000    Europa             55  

[319001 rows x 8 columns]

A continuación, añadimos el Elo estándar de cada jugador a finales de 2020.

In [66]:
df_id_2020 = pd.merge(df_players_cont,df_2020[df_2020.month == 12],on='fide_id',how = 'inner')
df_id_2020

fide_id                   name federation gender title   yob  \
0          100013     Campora, Daniel H.        ARG      M    GM  1957   
1          100021   Quinteros, Miguel A.        ARG      M    GM  1947   
2          100048     Iapichello, Javier        ARG      M   NaN  1990   
3          100064     Schipper, Bernardo        ARG      M   NaN  2000   
4          100072            Adla, Diego        ARG      M    IM  1968   
...           ...                    ...        ...    ...   ...   ...   
316483  651079542      Teisseire, Armand        FRA      M   NaN  2006   
316484  651080206         Lagarenne, Tom        FRA      M   NaN  1994   
316485  651081628        Kratz, Jonathan        FRA      M   NaN  1988   
316486  651081695          Bar, Francois        FRA      M   NaN  1965   
316487  651081830  Martinati, Jean Louis        FRA      M   NaN  1965   

       continent  years_id_2020  year  month  rating_standard  rating_rapid  \
0        América             63  2020     12           2420.0        2374.0   
1        América             73  2020     12           2422.0           NaN   
2        América             30  2020     12           1607.0           NaN   
3        América             20  2020     12           2116.0           NaN   
4        América             52  2020     12           2433.0           NaN   
...          ...            ...   ...    ...              ...           ...   
316483    Europa             14  2020     12              NaN        1473.0   
316484    Europa             26  2020     12           1327.0           NaN   
316485    Europa             32  2020     12           1195.0           NaN   
316486    Europa             55  2020     12              NaN           NaN   
316487    Europa             55  2020     12              NaN        1221.0   

        rating_blitz  
0             2378.0  
1                NaN  
2                NaN  
3                NaN  
4                NaN  
...              ...  
316483           NaN  
316484           NaN  
316485           NaN  
316486        1433.0  
316487           NaN  

[316488 rows x 13 columns]

In [67]:
df_id_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 316488 entries, 0 to 316487
Data columns (total 13 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   fide_id          316488 non-null  int64  
 1   name             316488 non-null  object 
 2   federation       316488 non-null  object 
 3   gender           316488 non-null  object 
 4   title            14139 non-null   object 
 5   yob              316488 non-null  int64  
 6   continent        316488 non-null  object 
 7   years_id_2020    316488 non-null  int64  
 8   year             316488 non-null  int64  
 9   month            316488 non-null  int64  
 10  rating_standard  266162 non-null  float64
 11  rating_rapid     149524 non-null  float64
 12  rating_blitz     105889 non-null  float64
dtypes: float64(3), int64(5), object(5)
memory usage: 31.4+ MB


Vemos que las columnas con nulos son Title y los tres ratings. 

Eliminemos aquellas columnas que no nos interesen:

In [68]:
df_id_2020 = df_id_2020.drop(columns=['rating_rapid', 'rating_blitz', 'year', 'month', 'yob'])
df_id_2020 = df_id_2020.rename(columns={'rating_standard':'rating_standard_2020'})
df_id_2020

fide_id                   name federation gender title continent  \
0          100013     Campora, Daniel H.        ARG      M    GM   América   
1          100021   Quinteros, Miguel A.        ARG      M    GM   América   
2          100048     Iapichello, Javier        ARG      M   NaN   América   
3          100064     Schipper, Bernardo        ARG      M   NaN   América   
4          100072            Adla, Diego        ARG      M    IM   América   
...           ...                    ...        ...    ...   ...       ...   
316483  651079542      Teisseire, Armand        FRA      M   NaN    Europa   
316484  651080206         Lagarenne, Tom        FRA      M   NaN    Europa   
316485  651081628        Kratz, Jonathan        FRA      M   NaN    Europa   
316486  651081695          Bar, Francois        FRA      M   NaN    Europa   
316487  651081830  Martinati, Jean Louis        FRA      M   NaN    Europa   

        years_id_2020  rating_standard_2020  
0                  63                2420.0  
1                  73                2422.0  
2                  30                1607.0  
3                  20                2116.0  
4                  52                2433.0  
...               ...                   ...  
316483             14                   NaN  
316484             26                1327.0  
316485             32                1195.0  
316486             55                   NaN  
316487             55                   NaN  

[316488 rows x 8 columns]

Añadimos rangos de edad, los cuales facilitarán análisis posteriores. 
Diremos lo siguiente: 
- De 0 a 8 años serán jugadores 'infant'
- De 9 a 18 años serán jugadores 'young'
- De 19 a 50 años serán jugadores 'adult'
- Superior a 50 años serán jugadores 'senior'


In [69]:
bins = [0, 8, 18, 50, float('inf')]  # Definimos los límites inferior y superior de cada categoría
labels = ['infant', 'young', 'adult', 'senior'] # Definimos las etiquetas

df_id_2020['category_age_2020'] = pd.cut(df_id_2020['years_id_2020'], bins=bins, labels=labels, right=False)
df_id_2020

fide_id                   name federation gender title continent  \
0          100013     Campora, Daniel H.        ARG      M    GM   América   
1          100021   Quinteros, Miguel A.        ARG      M    GM   América   
2          100048     Iapichello, Javier        ARG      M   NaN   América   
3          100064     Schipper, Bernardo        ARG      M   NaN   América   
4          100072            Adla, Diego        ARG      M    IM   América   
...           ...                    ...        ...    ...   ...       ...   
316483  651079542      Teisseire, Armand        FRA      M   NaN    Europa   
316484  651080206         Lagarenne, Tom        FRA      M   NaN    Europa   
316485  651081628        Kratz, Jonathan        FRA      M   NaN    Europa   
316486  651081695          Bar, Francois        FRA      M   NaN    Europa   
316487  651081830  Martinati, Jean Louis        FRA      M   NaN    Europa   

        years_id_2020  rating_standard_2020 category_age_2020  
0                  63                2420.0            senior  
1                  73                2422.0            senior  
2                  30                1607.0             adult  
3                  20                2116.0             adult  
4                  52                2433.0            senior  
...               ...                   ...               ...  
316483             14                   NaN             young  
316484             26                1327.0             adult  
316485             32                1195.0             adult  
316486             55                   NaN            senior  
316487             55                   NaN            senior  

[316488 rows x 9 columns]

In [70]:
df_id_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 316488 entries, 0 to 316487
Data columns (total 9 columns):
 #   Column                Non-Null Count   Dtype   
---  ------                --------------   -----   
 0   fide_id               316488 non-null  int64   
 1   name                  316488 non-null  object  
 2   federation            316488 non-null  object  
 3   gender                316488 non-null  object  
 4   title                 14139 non-null   object  
 5   continent             316488 non-null  object  
 6   years_id_2020         316488 non-null  int64   
 7   rating_standard_2020  266162 non-null  float64 
 8   category_age_2020     316488 non-null  category
dtypes: category(1), float64(1), int64(2), object(5)
memory usage: 19.6+ MB


Tenemos datos de todas las filas excepto de Title y de rating_standard_2020. 

Columna title:
Ahora bien, si analizamos más en profundidad el significado del NaN en esta variable, vemos que no es un valor que debamos sustituir o debamos eliminar. Nos está indicando simplemente que que hay jugadores que sí tienen un título y otros que no. Esto más adelante debe analizarse, pero no debe eliminarse porque puede darnos información muy interesante. 

Columna rating_standard_2020:
La mejor opción, al ser variables numéricas, es sustituirlo por la media. Ya que estamos preparando el dataset para analizar distinciones entre género, edad y país, para realizar la media deberían tenerse en cuenta estas variables.

In [71]:
medias = df_id_2020.groupby(['gender', 'category_age_2020', 'federation'])['rating_standard_2020'].mean()
medias

C:\Users\rjulianortin\AppData\Local\Temp\ipykernel_65336\2111246719.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  medias = df_id_2020.groupby(['gender', 'category_age_2020', 'federation'])['rating_standard_2020'].mean()


gender  category_age_2020  federation
F       infant             AFG                   NaN
                           ALB                   NaN
                           AND                   NaN
                           ANT                   NaN
                           ARG                   NaN
                                            ...     
M       senior             UKR           1969.816170
                           USA           1964.757299
                           UZB           1833.957447
                           VEN           1725.747253
                           YEM           1991.362069
Name: rating_standard_2020, Length: 904, dtype: float64

In [72]:
df_id_2020['rating_standard_2020'] = df_id_2020.apply(lambda row: medias[row['gender'], row['category_age_2020'], row['federation']] if pd.isna(row['rating_standard_2020']) else row['rating_standard_2020'], axis=1)
df_id_2020

fide_id                   name federation gender title continent  \
0          100013     Campora, Daniel H.        ARG      M    GM   América   
1          100021   Quinteros, Miguel A.        ARG      M    GM   América   
2          100048     Iapichello, Javier        ARG      M   NaN   América   
3          100064     Schipper, Bernardo        ARG      M   NaN   América   
4          100072            Adla, Diego        ARG      M    IM   América   
...           ...                    ...        ...    ...   ...       ...   
316483  651079542      Teisseire, Armand        FRA      M   NaN    Europa   
316484  651080206         Lagarenne, Tom        FRA      M   NaN    Europa   
316485  651081628        Kratz, Jonathan        FRA      M   NaN    Europa   
316486  651081695          Bar, Francois        FRA      M   NaN    Europa   
316487  651081830  Martinati, Jean Louis        FRA      M   NaN    Europa   

        years_id_2020  rating_standard_2020 category_age_2020  
0                  63           2420.000000            senior  
1                  73           2422.000000            senior  
2                  30           1607.000000             adult  
3                  20           2116.000000             adult  
4                  52           2433.000000            senior  
...               ...                   ...               ...  
316483             14           1302.864054             young  
316484             26           1327.000000             adult  
316485             32           1195.000000             adult  
316486             55           1607.055253            senior  
316487             55           1607.055253            senior  

[316488 rows x 9 columns]

In [73]:
df_id_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 316488 entries, 0 to 316487
Data columns (total 9 columns):
 #   Column                Non-Null Count   Dtype   
---  ------                --------------   -----   
 0   fide_id               316488 non-null  int64   
 1   name                  316488 non-null  object  
 2   federation            316488 non-null  object  
 3   gender                316488 non-null  object  
 4   title                 14139 non-null   object  
 5   continent             316488 non-null  object  
 6   years_id_2020         316488 non-null  int64   
 7   rating_standard_2020  316441 non-null  float64 
 8   category_age_2020     316488 non-null  category
dtypes: category(1), float64(1), int64(2), object(5)
memory usage: 19.6+ MB


A pesar de haber rellenado hasta el Elo FIDE de más de 50.000 jugadores, todavía siguen existiendo nulos en rating_standard_2020. Lo mas probable es que se deba a jugadores que no tienen más referentes en país - edad - género. 

Para que no afecten en nuestros análisis, eliminaremos estas filas (un total de 47 filas)

In [74]:
df_sin_nulls = df_id_2020.copy()
df_sin_nulls = df_sin_nulls.dropna(subset=['rating_standard_2020'])
df_sin_nulls.info()

<class 'pandas.core.frame.DataFrame'>
Index: 316441 entries, 0 to 316487
Data columns (total 9 columns):
 #   Column                Non-Null Count   Dtype   
---  ------                --------------   -----   
 0   fide_id               316441 non-null  int64   
 1   name                  316441 non-null  object  
 2   federation            316441 non-null  object  
 3   gender                316441 non-null  object  
 4   title                 14138 non-null   object  
 5   continent             316441 non-null  object  
 6   years_id_2020         316441 non-null  int64   
 7   rating_standard_2020  316441 non-null  float64 
 8   category_age_2020     316441 non-null  category
dtypes: category(1), float64(1), int64(2), object(5)
memory usage: 22.0+ MB


El Elo FIDE no tiene decimales, de modo que los quitaremos:

In [75]:
df_sin_nulls['rating_standard_2020'] = df_sin_nulls['rating_standard_2020'].astype(int)
df_sin_nulls

fide_id                   name federation gender title continent  \
0          100013     Campora, Daniel H.        ARG      M    GM   América   
1          100021   Quinteros, Miguel A.        ARG      M    GM   América   
2          100048     Iapichello, Javier        ARG      M   NaN   América   
3          100064     Schipper, Bernardo        ARG      M   NaN   América   
4          100072            Adla, Diego        ARG      M    IM   América   
...           ...                    ...        ...    ...   ...       ...   
316483  651079542      Teisseire, Armand        FRA      M   NaN    Europa   
316484  651080206         Lagarenne, Tom        FRA      M   NaN    Europa   
316485  651081628        Kratz, Jonathan        FRA      M   NaN    Europa   
316486  651081695          Bar, Francois        FRA      M   NaN    Europa   
316487  651081830  Martinati, Jean Louis        FRA      M   NaN    Europa   

        years_id_2020  rating_standard_2020 category_age_2020  
0                  63                  2420            senior  
1                  73                  2422            senior  
2                  30                  1607             adult  
3                  20                  2116             adult  
4                  52                  2433            senior  
...               ...                   ...               ...  
316483             14                  1302             young  
316484             26                  1327             adult  
316485             32                  1195             adult  
316486             55                  1607            senior  
316487             55                  1607            senior  

[316441 rows x 9 columns]

Para que se detecten las personas que no tienen título, cambiaremos el valor de NaN por No title.

In [76]:
df_sin_nulls['title'] = df_sin_nulls['title'].fillna('No title')
df_sin_nulls

fide_id                   name federation gender     title  \
0          100013     Campora, Daniel H.        ARG      M        GM   
1          100021   Quinteros, Miguel A.        ARG      M        GM   
2          100048     Iapichello, Javier        ARG      M  No title   
3          100064     Schipper, Bernardo        ARG      M  No title   
4          100072            Adla, Diego        ARG      M        IM   
...           ...                    ...        ...    ...       ...   
316483  651079542      Teisseire, Armand        FRA      M  No title   
316484  651080206         Lagarenne, Tom        FRA      M  No title   
316485  651081628        Kratz, Jonathan        FRA      M  No title   
316486  651081695          Bar, Francois        FRA      M  No title   
316487  651081830  Martinati, Jean Louis        FRA      M  No title   

       continent  years_id_2020  rating_standard_2020 category_age_2020  
0        América             63                  2420            senior  
1        América             73                  2422            senior  
2        América             30                  1607             adult  
3        América             20                  2116             adult  
4        América             52                  2433            senior  
...          ...            ...                   ...               ...  
316483    Europa             14                  1302             young  
316484    Europa             26                  1327             adult  
316485    Europa             32                  1195             adult  
316486    Europa             55                  1607            senior  
316487    Europa             55                  1607            senior  

[316441 rows x 9 columns]

Hemos encontrado casos en que years_id_2020 es superior a 200 años. Esto se debe a que el dato de yob era de valor 0. Limpiemos esas filas.

In [78]:
df_sin_nulls = df_sin_nulls.drop(df_sin_nulls[df_sin_nulls['years_id_2020'] == 2020].index)

In [79]:
df_sin_nulls

fide_id                   name federation gender     title  \
0          100013     Campora, Daniel H.        ARG      M        GM   
1          100021   Quinteros, Miguel A.        ARG      M        GM   
2          100048     Iapichello, Javier        ARG      M  No title   
3          100064     Schipper, Bernardo        ARG      M  No title   
4          100072            Adla, Diego        ARG      M        IM   
...           ...                    ...        ...    ...       ...   
316483  651079542      Teisseire, Armand        FRA      M  No title   
316484  651080206         Lagarenne, Tom        FRA      M  No title   
316485  651081628        Kratz, Jonathan        FRA      M  No title   
316486  651081695          Bar, Francois        FRA      M  No title   
316487  651081830  Martinati, Jean Louis        FRA      M  No title   

       continent  years_id_2020  rating_standard_2020 category_age_2020  
0        América             63                  2420            senior  
1        América             73                  2422            senior  
2        América             30                  1607             adult  
3        América             20                  2116             adult  
4        América             52                  2433            senior  
...          ...            ...                   ...               ...  
316483    Europa             14                  1302             young  
316484    Europa             26                  1327             adult  
316485    Europa             32                  1195             adult  
316486    Europa             55                  1607            senior  
316487    Europa             55                  1607            senior  

[311270 rows x 9 columns]

In [81]:
df_sin_nulls.title.value_counts()

title
No title    297208
FM            5877
IM            2911
GM            1343
WFM           1323
CM            1230
WIM            605
WCM            494
WGM            254
NI              15
DI               7
FI               3
Name: count, dtype: int64

In [82]:
df_sin_nulls[df_sin_nulls.title == 'NI'] #vemos que hay categorías que sí distinguen por F o M, pero hay otras que no. 

fide_id                              name federation gender title  \
68698     1950720                   Spinola, Sidney        CPV      M    NI   
71436     2051400               Mc Roberts, Holly J        USA      F    NI   
71900     2066971                    Schley, Andrew        USA      M    NI   
72722     2093235              Wyzywany, Annastasia        USA      F    NI   
105530    3830470  Murillo Pittman, Oswaldo Segundo        PER      M    NI   
143410    7608551                     Jaafar,Issa .        SYR      M    NI   
147637   10005013             Mande, Allan Benjamin        UGA      M    NI   
151850   10806873                  Michuki, Francis        KEN      M    NI   
162513   13708341                    Kvitko, Nikita        KAZ      M    NI   
165604   13800442              Derkembaev, Samatbek        KGZ      M    NI   
165755   13803239             Tilenbaev, Cholponbek        KGZ      M    NI   
176654   14801558                Cherene, Joao Jefu        MOZ      M    NI   
181501   15205681                Nghifikwa, Leonard        NAM      M    NI   
181769   16401131    Metuno, Kuete Bleriot Rocalixt        CMR      M    NI   
313323  381043108                   Damir, Talapker        KGZ      M    NI   

       continent  years_id_2020  rating_standard_2020 category_age_2020  
68698     África             41                  1663             adult  
71436    América             44                  1702             adult  
71900    América             27                  1936             adult  
72722    América             23                  1654             adult  
105530   América             55                  1229            senior  
143410      Asia             27                  1767             adult  
147637    África             27                  1874             adult  
151850    África             23                  1637             adult  
162513      Asia             26                  1635             adult  
165604      Asia             40                  1701             adult  
165755      Asia             65                  1683            senior  
176654    África             39                  1838             adult  
181501    África             28                  1680             adult  
181769    África             24                  1829             adult  
313323      Asia             25                  1315             adult

Si analizamos las variables dentro de title, vemos que hay cosas tales como WGM (Woman Grand Master) y GM (Grand Master), es decir, distingue entre entre género F y M. Pero no siempre es así. Cambiaremos aquellas que empiezan por W y las categorizaremos dentro del mismo grupo.

In [83]:
categorias_a_renombrar = {'WFM': 'FM',
                          'WIM': 'IM',
                          'WCM': 'CM',
                          'WGM': 'GM',
                         }

df_2020_title_renamed = df_sin_nulls.copy()
df_2020_title_renamed['title'] = df_2020_title_renamed['title'].replace(categorias_a_renombrar)
df_2020_title_renamed.title.value_counts()

title
No title    297208
FM            7200
IM            3516
CM            1724
GM            1597
NI              15
DI               7
FI               3
Name: count, dtype: int64

Describimos a continuación todas las columnas de este nuevo dataset:

| Columna/Variable | Descripción |
| ---------------- | ----------- |
| fide_id         | Número de identificación FIDE |
| name           | Nombre del jugador o jugadora |
| federation              | País del jugador o jugadora |
| gender              | Jugador de género masculino (M) o femenino (F) |
| title            | Titulo |
| continent            | Continente del jugador o jugadora |
| years_id_2020             | Edad del ajedrecista en el año 2020 |
| rating_standard_2020            | ELO estándar del año 2020 |
| category_age_2020            | Categoría de edad en el año 2020 |

In [84]:
#Guardamos el dataset
df_2020_title_renamed.to_csv('datasets_clean/players_clean.csv', index=False)  

### Dataset 2: Unión de los datasets desde enero 2015 hasta diciembre 2020

In [86]:
df_2015_2020 = pd.concat([df_2015, df_2016, df_2017, df_2018, df_2019, df_2020], ignore_index=True)
df_2015_2020 = df_2015_2020.drop(columns= ['rating_rapid','rating_blitz'])
df_2015_2020 = pd.DataFrame(df_2015_2020)
df_2015_2020

fide_id  year  month  rating_standard
0            100013  2015      1           2456.0
1            100021  2015      1           2422.0
2            100048  2015      1           1607.0
3            100064  2015      1           2116.0
4            100072  2015      1           2469.0
...             ...   ...    ...              ...
23700033  651079542  2020     12              NaN
23700034  651080206  2020     12           1327.0
23700035  651081628  2020     12           1195.0
23700036  651081695  2020     12              NaN
23700037  651081830  2020     12              NaN

[23700038 rows x 4 columns]

In [87]:
cantidad_nulos = df_2015_2020['rating_standard'].isnull().sum()
cantidad_nulos

2713647

Observamos una gran cantidad de nulos en la columna rating_standard.
Siguiendo el mismo procedimiento anteriormente mencionado, procederemos a sustituirlos por la media distinguiendo entre género, edad y país. 

In [88]:
df_2015_2020_merged = pd.merge(df_2015_2020, df_players, on='fide_id',how = 'inner')

In [89]:
df_2015_2020_merged

fide_id  year  month  rating_standard                   name  \
0            100013  2015      1           2456.0     Campora, Daniel H.   
1            100021  2015      1           2422.0   Quinteros, Miguel A.   
2            100048  2015      1           1607.0     Iapichello, Javier   
3            100064  2015      1           2116.0     Schipper, Bernardo   
4            100072  2015      1           2469.0            Adla, Diego   
...             ...   ...    ...              ...                    ...   
23605208  651079542  2020     12              NaN      Teisseire, Armand   
23605209  651080206  2020     12           1327.0         Lagarenne, Tom   
23605210  651081628  2020     12           1195.0        Kratz, Jonathan   
23605211  651081695  2020     12              NaN          Bar, Francois   
23605212  651081830  2020     12              NaN  Martinati, Jean Louis   

         federation gender title   yob  
0               ARG      M    GM  1957  
1               ARG      M    GM  1947  
2               ARG      M   NaN  1990  
3               ARG      M   NaN  2000  
4               ARG      M    IM  1968  
...             ...    ...   ...   ...  
23605208        FRA      M   NaN  2006  
23605209        FRA      M   NaN  1994  
23605210        FRA      M   NaN  1988  
23605211        FRA      M   NaN  1965  
23605212        FRA      M   NaN  1965  

[23605213 rows x 9 columns]

In [90]:
df_2015_2020_merged['years_id'] = df_2015_2020_merged['year'] - df_2015_2020_merged['yob']
df_2015_2020_merged['category_age'] = pd.cut(df_2015_2020_merged['years_id'], bins=bins, labels=labels, right=False) #reaprovechamos formula
df_2015_2020_merged

fide_id  year  month  rating_standard                   name  \
0            100013  2015      1           2456.0     Campora, Daniel H.   
1            100021  2015      1           2422.0   Quinteros, Miguel A.   
2            100048  2015      1           1607.0     Iapichello, Javier   
3            100064  2015      1           2116.0     Schipper, Bernardo   
4            100072  2015      1           2469.0            Adla, Diego   
...             ...   ...    ...              ...                    ...   
23605208  651079542  2020     12              NaN      Teisseire, Armand   
23605209  651080206  2020     12           1327.0         Lagarenne, Tom   
23605210  651081628  2020     12           1195.0        Kratz, Jonathan   
23605211  651081695  2020     12              NaN          Bar, Francois   
23605212  651081830  2020     12              NaN  Martinati, Jean Louis   

         federation gender title   yob  years_id category_age  
0               ARG      M    GM  1957        58       senior  
1               ARG      M    GM  1947        68       senior  
2               ARG      M   NaN  1990        25        adult  
3               ARG      M   NaN  2000        15        young  
4               ARG      M    IM  1968        47        adult  
...             ...    ...   ...   ...       ...          ...  
23605208        FRA      M   NaN  2006        14        young  
23605209        FRA      M   NaN  1994        26        adult  
23605210        FRA      M   NaN  1988        32        adult  
23605211        FRA      M   NaN  1965        55       senior  
23605212        FRA      M   NaN  1965        55       senior  

[23605213 rows x 11 columns]

In [91]:
medias_2015_2020 = df_2015_2020_merged.groupby(['gender', 'category_age', 'federation'])['rating_standard'].mean()
medias_2015_2020

C:\Users\rjulianortin\AppData\Local\Temp\ipykernel_65336\2466389858.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  medias_2015_2020 = df_2015_2020_merged.groupby(['gender', 'category_age', 'federation'])['rating_standard'].mean()


gender  category_age  federation
F       infant        AFG                   NaN
                      AHO                   NaN
                      ALB                   NaN
                      ALG           1056.000000
                      AND                   NaN
                                       ...     
M       senior        VIE           1895.547059
                      WLS           1879.896816
                      YEM           2015.877406
                      ZAM           1849.585436
                      ZIM           2033.738372
Name: rating_standard, Length: 1544, dtype: float64

In [92]:
df_2015_2020_merged['rating_standard'] = df_2015_2020_merged.apply(lambda row: medias_2015_2020[row['gender'], row['category_age'], row['federation']] if pd.isna(row['rating_standard']) else row['rating_standard'], axis=1)
df_2015_2020_merged

fide_id  year  month  rating_standard                   name  \
0            100013  2015      1      2456.000000     Campora, Daniel H.   
1            100021  2015      1      2422.000000   Quinteros, Miguel A.   
2            100048  2015      1      1607.000000     Iapichello, Javier   
3            100064  2015      1      2116.000000     Schipper, Bernardo   
4            100072  2015      1      2469.000000            Adla, Diego   
...             ...   ...    ...              ...                    ...   
23605208  651079542  2020     12      1359.843204      Teisseire, Armand   
23605209  651080206  2020     12      1327.000000         Lagarenne, Tom   
23605210  651081628  2020     12      1195.000000        Kratz, Jonathan   
23605211  651081695  2020     12      1646.705577          Bar, Francois   
23605212  651081830  2020     12      1646.705577  Martinati, Jean Louis   

         federation gender title   yob  years_id category_age  
0               ARG      M    GM  1957        58       senior  
1               ARG      M    GM  1947        68       senior  
2               ARG      M   NaN  1990        25        adult  
3               ARG      M   NaN  2000        15        young  
4               ARG      M    IM  1968        47        adult  
...             ...    ...   ...   ...       ...          ...  
23605208        FRA      M   NaN  2006        14        young  
23605209        FRA      M   NaN  1994        26        adult  
23605210        FRA      M   NaN  1988        32        adult  
23605211        FRA      M   NaN  1965        55       senior  
23605212        FRA      M   NaN  1965        55       senior  

[23605213 rows x 11 columns]

In [93]:
cantidad_nulos_2015_2020 = df_2015_2020_merged['rating_standard'].isnull().sum()
cantidad_nulos_2015_2020

637

Eliminaremos los 637 nulos restantes. Estos nulos se deben a que no hay una relación posible entre género, país y edad.

In [94]:
df_sin_nulls_2015_2020 = df_2015_2020_merged.copy()
df_sin_nulls_2015_2020 = df_sin_nulls_2015_2020.dropna(subset=['rating_standard'])
df_sin_nulls_2015_2020.info()

<class 'pandas.core.frame.DataFrame'>
Index: 23604576 entries, 0 to 23605212
Data columns (total 11 columns):
 #   Column           Dtype   
---  ------           -----   
 0   fide_id          int64   
 1   year             int64   
 2   month            int64   
 3   rating_standard  float64 
 4   name             object  
 5   federation       object  
 6   gender           object  
 7   title            object  
 8   yob              int64   
 9   years_id         int64   
 10  category_age     category
dtypes: category(1), float64(1), int64(5), object(4)
memory usage: 2.0+ GB


El Elo FIDE no tiene decimales:

In [95]:
df_sin_nulls_2015_2020['rating_standard'] = df_sin_nulls_2015_2020['rating_standard'].astype(int)
df_sin_nulls_2015_2020

fide_id  year  month  rating_standard                   name  \
0            100013  2015      1             2456     Campora, Daniel H.   
1            100021  2015      1             2422   Quinteros, Miguel A.   
2            100048  2015      1             1607     Iapichello, Javier   
3            100064  2015      1             2116     Schipper, Bernardo   
4            100072  2015      1             2469            Adla, Diego   
...             ...   ...    ...              ...                    ...   
23605208  651079542  2020     12             1359      Teisseire, Armand   
23605209  651080206  2020     12             1327         Lagarenne, Tom   
23605210  651081628  2020     12             1195        Kratz, Jonathan   
23605211  651081695  2020     12             1646          Bar, Francois   
23605212  651081830  2020     12             1646  Martinati, Jean Louis   

         federation gender title   yob  years_id category_age  
0               ARG      M    GM  1957        58       senior  
1               ARG      M    GM  1947        68       senior  
2               ARG      M   NaN  1990        25        adult  
3               ARG      M   NaN  2000        15        young  
4               ARG      M    IM  1968        47        adult  
...             ...    ...   ...   ...       ...          ...  
23605208        FRA      M   NaN  2006        14        young  
23605209        FRA      M   NaN  1994        26        adult  
23605210        FRA      M   NaN  1988        32        adult  
23605211        FRA      M   NaN  1965        55       senior  
23605212        FRA      M   NaN  1965        55       senior  

[23604576 rows x 11 columns]

In [96]:
df = df_sin_nulls_2015_2020.copy()
df = pd.merge(df_paises_ISO_fed,df,on='federation',how = 'inner')
df

federation continent   fide_id  year  month  rating_standard  \
0               AFG      Asia   1025511  2015      1             1528   
1               AFG      Asia  11700017  2015      1             1816   
2               AFG      Asia  11700025  2015      1             2000   
3               AFG      Asia  11700033  2015      1             1987   
4               AFG      Asia  11700041  2015      1             1861   
...             ...       ...       ...   ...    ...              ...   
17444112        DJI    África  32500971  2020     12             1872   
17444113        DJI    África  32500980  2020     12             1322   
17444114        DJI    África  32501013  2020     12             1824   
17444115        DJI    África  32501021  2020     12             1970   
17444116        DJI    África  32501048  2020     12             1473   

                              name gender title   yob  years_id category_age  
0                   Kamal, Mustafa      M   NaN  1962        53       senior  
1                Nuristani, A. Dr.      M   NaN  1933        82       senior  
2              Asefi, Zaheeruddeen      M    FM  1965        50       senior  
3                Ahmadi, Zabiullah      M   NaN  1988        27        adult  
4             Baryalai, Salahuddin      M   NaN  1962        53       senior  
...                            ...    ...   ...   ...       ...          ...  
17444112   Abdilahi Salah, Mohamed      M   NaN  1985        35        adult  
17444113    Adam Bouni, Abdi Rahim      M   NaN  1990        30        adult  
17444114      Osman Dahir, Mahamed      M   NaN  1994        26        adult  
17444115      Yousuf Omer, Mahamed      M   NaN  1994        26        adult  
17444116  Ismael, Mohamed Goumaneh      M   NaN  1983        37        adult  

[17444117 rows x 12 columns]

In [97]:
df = df.drop(columns = ['yob','name'])
df

federation continent   fide_id  year  month  rating_standard gender  \
0               AFG      Asia   1025511  2015      1             1528      M   
1               AFG      Asia  11700017  2015      1             1816      M   
2               AFG      Asia  11700025  2015      1             2000      M   
3               AFG      Asia  11700033  2015      1             1987      M   
4               AFG      Asia  11700041  2015      1             1861      M   
...             ...       ...       ...   ...    ...              ...    ...   
17444112        DJI    África  32500971  2020     12             1872      M   
17444113        DJI    África  32500980  2020     12             1322      M   
17444114        DJI    África  32501013  2020     12             1824      M   
17444115        DJI    África  32501021  2020     12             1970      M   
17444116        DJI    África  32501048  2020     12             1473      M   

         title  years_id category_age  
0          NaN        53       senior  
1          NaN        82       senior  
2           FM        50       senior  
3          NaN        27        adult  
4          NaN        53       senior  
...        ...       ...          ...  
17444112   NaN        35        adult  
17444113   NaN        30        adult  
17444114   NaN        26        adult  
17444115   NaN        26        adult  
17444116   NaN        37        adult  

[17444117 rows x 10 columns]

In [98]:
df['title'] = df['title'].fillna('No title')

In [101]:
df = df.loc[df['years_id'] <= 120]
df

federation continent   fide_id  year  month  rating_standard gender  \
0               AFG      Asia   1025511  2015      1             1528      M   
1               AFG      Asia  11700017  2015      1             1816      M   
2               AFG      Asia  11700025  2015      1             2000      M   
3               AFG      Asia  11700033  2015      1             1987      M   
4               AFG      Asia  11700041  2015      1             1861      M   
...             ...       ...       ...   ...    ...              ...    ...   
17444112        DJI    África  32500971  2020     12             1872      M   
17444113        DJI    África  32500980  2020     12             1322      M   
17444114        DJI    África  32501013  2020     12             1824      M   
17444115        DJI    África  32501021  2020     12             1970      M   
17444116        DJI    África  32501048  2020     12             1473      M   

             title  years_id category_age  
0         No title        53       senior  
1         No title        82       senior  
2               FM        50       senior  
3         No title        27        adult  
4         No title        53       senior  
...            ...       ...          ...  
17444112  No title        35        adult  
17444113  No title        30        adult  
17444114  No title        26        adult  
17444115  No title        26        adult  
17444116  No title        37        adult  

[17100755 rows x 10 columns]

In [103]:
df.title.value_counts()

title
No title    16098805
FM            421305
IM            209464
GM             96684
WFM            94465
CM             84969
WIM            43536
WCM            32184
WGM            18288
NI               654
FI               216
DI               185
Name: count, dtype: int64

Describimos a continuación todas las columnas de este nuevo dataset:

| Columna/Variable | Descripción |
| ---------------- | ----------- |
| federation            | País del jugador o jugadora |
| continent            | Continente del jugador o jugadora |
| fide_id         | Número de identificación FIDE |
| year          | Año |
| month              | Mes |
| rating_standard           | ELO estándar del año 2020 |
| name           | Nombre del jugador o jugadora | Discreta |
| gender              | Jugador de género masculino (M) o femenino (F) |
| title            | Titulo | 
| years_id             | Edad del ajedrecista en el año indicado en la columna year | 
| category_age            | Categoría de edad en el año indicado en la columna year| 


In [102]:
#Guardamos el dataset
df.to_csv('datasets_clean/evolution_clean.csv', index=False)  